File: src/util/graphics/IconExtractor.java
package util.graphics;

import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import javafx.embed.swing.SwingFXUtils;
import javafx.scene.image.Image;
import javax.swing.filechooser.FileSystemView;
import util.LazyR;
import util.R;
import util.file.Util;
import util.file.WindowsShortcut;
import static java.awt.image.BufferedImage.TYPE_INT_ARGB;
import static util.file.UtilKt.getNameWithoutExtensionOrRoot;

/**
 * Extracts an icon for a file type of specific file.
 * <p/>
 * http://stackoverflow.com/questions/15629069/extract-application-icons-from-desktop-folder-to-application
 * <br/>
 * http://stackoverflow.com/questions/15149565/how-does-jtree-display-file-name/15150756#15150756
 * <br/>
 * http://stackoverflow.com/questions/28034432/javafx-file-listview-with-icon-and-file-name
 * <br/>
 * http://stackoverflow.com/questions/26192832/java-javafx-set-swing-icon-for-javafx-label
 */
public class IconExtractor {

	private static final R<FileSystemView> helperFileSystemView = new LazyR<>(FileSystemView::getFileSystemView);
	private static final Map<String,Image> mapOfFileExtToSmallIcon = new ConcurrentHashMap<>();

	private static javax.swing.Icon getSwingIconFromFileSystem(File file) {

		// Windows
		return helperFileSystemView.get().getSystemIcon(file);

		// OS X
		//final javax.swing.JFileChooser fc = new javax.swing.JFileChooser();
		//return icon = fc.getUI().getFileView(fc).getIcon(file);
	}

	public static Image getFileIcon(File file) {
		String ext = Util.getSuffix(file.getPath()).toLowerCase();

		// Handle windows shortcut files (we need to resolve the target file)
		if ("lnk".equals(ext))
			return WindowsShortcut.targetedFile(file).map(IconExtractor::getFileIcon).orElse(null);

		// Handle windows executable files (we need to handle each individually)
		String key = "exe".equals(ext) ? getNameWithoutExtensionOrRoot(file) : ext;

		return mapOfFileExtToSmallIcon.computeIfAbsent(key, k -> {
			javax.swing.Icon swingIcon = null;
			if (file.exists()) {
				swingIcon = getSwingIconFromFileSystem(file);
			} else {
				File tempFile = null;
				try {
					tempFile = File.createTempFile("icon", ext);
					swingIcon = getSwingIconFromFileSystem(tempFile);
				} catch (IOException ignored) {
					// Cannot create temporary file.
				} finally {
					if (tempFile!=null) tempFile.delete();
				}
			}
			return swingIconToImage(swingIcon);
		});
	}

	private static Image swingIconToImage(javax.swing.Icon swingIcon) {
		if (swingIcon==null) return null;
		BufferedImage image = new BufferedImage(swingIcon.getIconWidth(), swingIcon.getIconHeight(), TYPE_INT_ARGB);
		swingIcon.paintIcon(null, image.getGraphics(), 0, 0);
		return SwingFXUtils.toFXImage(image, null);
	}

}

File: src/sp/it/pl/util/graphics/IconExtractor.kt
package sp.it.pl.util.graphics

import javafx.scene.image.Image
import sp.it.pl.util.file.Util
import sp.it.pl.util.file.WindowsShortcut
import sp.it.pl.util.file.div
import sp.it.pl.util.file.mimetype.MimeExt.Companion.exe
import sp.it.pl.util.file.mimetype.MimeExt.Companion.lnk
import sp.it.pl.util.file.nameWithoutExtensionOrRoot
import sp.it.pl.util.functional.orNull
import sp.it.pl.util.functional.runIf
import sp.it.pl.util.graphics.image.toFX
import sp.it.pl.util.system.Os
import java.awt.image.BufferedImage
import java.awt.image.BufferedImage.TYPE_INT_ARGB
import java.io.File
import java.io.IOException
import java.util.concurrent.ConcurrentHashMap
import javax.swing.Icon
import javax.swing.filechooser.FileSystemView

/**
 * Extracts an icon for a file type of specific file.
 *
 * http://stackoverflow.com/questions/15629069/extract-application-icons-from-desktop-folder-to-application
 * http://stackoverflow.com/questions/15149565/how-does-jtree-display-file-name/15150756#15150756
 * http://stackoverflow.com/questions/28034432/javafx-file-listview-with-icon-and-file-name
 * http://stackoverflow.com/questions/26192832/java-javafx-set-swing-icon-for-javafx-label
 */
object IconExtractor {
    private val dirTmp = File(System.getProperty("java.io.tmpdir"))
    private val helperFileSystemView by lazy { FileSystemView.getFileSystemView() }
    private val mapOfFileExtToSmallIcon = ConcurrentHashMap<String, Image?>()

    @JvmStatic fun getFileIcon(file: File): Image? {

        val ext = Util.getSuffix(file.path).toLowerCase()

        // shortcuts have icons of files they refer to
        if (lnk==ext)
            return WindowsShortcut.targetedFile(file).map(::getFileIcon).orNull()

        // Handle windows executable files (we need to handle each individually)
        val isExe = exe==ext
        val key = if (isExe) file.nameWithoutExtensionOrRoot else ext

        return mapOfFileExtToSmallIcon.computeIfAbsent(key) {
            val iconFile = file.takeIf { it.exists() } ?: runIf(!isExe) {
                try {
                    val f = dirTmp/"file_type_icons.$it"
                    if (!f.exists()) f.writeText("")
                    f
                } catch (ignored: IOException) {
                    null
                }
            }
            iconFile?.getSwingIconFromFileSystem()?.toImage()
        }
    }

    private fun File.getSwingIconFromFileSystem(): Icon? = when (Os.current) {
        Os.WINDOWS -> helperFileSystemView.getSystemIcon(this)
        // TODO: implement
        // Os.OSX -> {
        //     final javax.swing.JFileChooser fc = new javax.swing.JFileChooser();
        //     return icon = fc.getUI().getFileView(fc).getIcon(file);
        // }
        else -> null
    }

    private fun Icon.toImage(): Image? {
        val image = BufferedImage(this.iconWidth, this.iconHeight, TYPE_INT_ARGB)
        paintIcon(null, image.graphics, 0, 0)
        return image.toFX(null)
    }

}